data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java

```diff
@@ -0,0 +1,296 @@
+            jsonOutputCodecConfig.setKeyName(batchOptions.getKeyName());
+            codec = new JsonOutputCodec(jsonOutputCodecConfig);
+            maxEvents = batchOptions.getThresholdOptions().getEventCount();
+            maxBytes = batchOptions.getThresholdOptions().getMaximumSize(); // remove
+            maxCollectionDuration = batchOptions.getThresholdOptions().getEventCollectTimeOut();
+            isBatchEnabled = true;
+            LOG.info("maxEvents:" + maxEvents + " maxbytes:" + maxBytes + " maxDuration:" + maxCollectionDuration);
+        } else if(payloadModel.equals(SINGLE_EVENT)) {
+            NdjsonOutputConfig ndjsonOutputCodecConfig = new NdjsonOutputConfig();
+            codec = new NdjsonOutputCodec(ndjsonOutputCodecConfig);
+            isBatchEnabled = false;
+        } else{
+            throw new RuntimeException("invalid payload_model option");
+        }
+
+        if(!lambdaProcessorConfig.getInvocationType().equals(LambdaCommonConfig.EVENT) &&
+                !lambdaProcessorConfig.getInvocationType().equals(LambdaCommonConfig.REQUEST_RESPONSE)){
+            throw new RuntimeException("Unsupported invocation type " + lambdaProcessorConfig.getInvocationType());
+        }
+
+        //Translate dataprepper invocation type to lambda invocation type
+        Map<String, String> invocationTypeMap = Map.of(
+                LambdaCommonConfig.EVENT, EVENT_LAMBDA,
+                LambdaCommonConfig.REQUEST_RESPONSE, REQUEST_RESPONSE_LAMBDA
+        );
+        invocationType = invocationTypeMap.get(lambdaProcessorConfig.getInvocationType());
+
+        bufferedEventHandles = new LinkedList<>();
+        events = new ArrayList();
+
+        lambdaClient = LambdaClientFactory.createLambdaClient(lambdaProcessorConfig.getAwsAuthenticationOptions(),
+                lambdaProcessorConfig.getMaxConnectionRetries()
+                , awsCredentialsSupplier);
+
+        this.bufferFactory = new InMemoryBufferFactory();
+        try {
+            currentBuffer = this.bufferFactory.getBuffer(lambdaClient, functionName, invocationType);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    @Override
+    public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
+        if (records.isEmpty()) {
+            return records;
+        }
+
+        //lambda mutates event
+        List<Record<Event>> resultRecords = new ArrayList<>();
+
+        for (Record<Event> record : records) {
+            final Event event = record.getData();
+
+            if (whenCondition != null && !expressionEvaluator.evaluateConditional(whenCondition, event)) {
+                continue;
+            }
+
+            try {
+                if (currentBuffer.getEventCount() == 0) {
+                    codec.start(currentBuffer.getOutputStream(), event, codecContext);
+                }
+                codec.writeEvent(event, currentBuffer.getOutputStream());
```

Comment:
is `currentBuffer` thread safe ?

Reply:
yes this is running in the context of a processor.